api/src/main/java/org/eclipse/microprofile/metrics/DefaultGlobalTagsProvider.java

```diff
@@ -0,0 +1,135 @@
+/*
+ **********************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICES file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ **********************************************************************/
+package org.eclipse.microprofile.metrics;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+
+import org.eclipse.microprofile.config.Config;
+import org.eclipse.microprofile.config.ConfigProvider;
+
+/**
+ * Default implementation of {@link org.eclipse.microprofile.metrics.GlobalTagsProvider} SPI.
+ * <p>
+ * This class is intentionally package private, and NOT discoverable by the
+ * {@code ServiceLoader}, as it is intended to be used only by the exiting
+ * {@link org.eclipse.microprofile.metrics.MetricID#MetricID(String, Tag...)
+ * MetricID constructor}, in order to preserve backwards compatibility.
+ * </p>
+ * <p>
+ * It is recommended that the MP Metrics implementations provide their own
+ * implementation of this behavior.
+ * </p>
+ */
+class DefaultGlobalTagsProvider implements GlobalTagsProvider {
+    // singleton instance that caches the results of the global tags lookup
+    static final GlobalTagsProvider INSTANCE = new DefaultGlobalTagsProvider();
+
+    private static final String GLOBAL_TAGS_VARIABLE = "mp.metrics.tags";
+
+    private static final String APPLICATION_NAME_VARIABLE = "mp.metrics.appName";
+    private static final String APPLICATION_NAME_TAG = "_app";
+
+    private static final String GLOBAL_TAG_MALFORMED_EXCEPTION = "Malformed list of Global Tags. Tag names "
+                                                                + "must match the following regex [a-zA-Z_][a-zA-Z0-9_]*."
+                                                                + " Global Tag values must not be empty."
+                                                                + " Global Tag values MUST escape equal signs `=` and commas `,`"
+                                                                + " with a backslash `\\` ";
+
+    private volatile Set<Tag> tags = null;
```

Comment:
I think I see a problem for application servers with multiple applications here. If the `DefaultGlobalTagsProvider` provider is a singleton and it caches the global tags once, there is no way we can get different tags for each application in case that all applications share the same singleton instance, which they will if they use a shared class loader to load MP Metrics classes.

So how could we differentiate between apps? Perhaps we could solve this by not caching the returned tags from providers found by ServiceLoader, and instead letting the custom providers do the caching themselves? Each time `getGlobalTags` is invoked, the custom providers would need to be called again. And the custom provider would know what application it is serving so it could return the right `_app` tag. We'd also need to make sure that `_app` tags from the custom provider aren't overwritten by the `DefaultGlobalTagsProvider`.

Reply:
> Perhaps we could solve this by not caching the returned tags from providers found by ServiceLoader, and instead letting the custom providers do the caching themselves?

That sounds like a good approach.  Would be good to see different servers try this to see if it works well for single and multi app servers.

Reply:
@jmartisk The default implementation is meant to be just that -- the default.

I believe it is reasonable to expect app servers to provide their own implementation, which takes all these things you mentioned (multiple apps, using multiple class loaders) into account and pass it to the new 3-arg constructor when creating `MetricIDs`.

As a matter of fact, I think everyone should do that, including frameworks that are not app servers and don't support multiple apps in the same process. The default implementation is there purely for compatibility reasons, to avoid breaking anyone right away. Maybe we should even deprecate the 2-arg constructor, so everyone knows they should not use it long term.

Reply:
@jmartisk I moved the singleton instance that caches global tags lookup from `DefaultGlobalTagsProvider` to `MetricID`.

This allows `DefaulGlobalTagsProvider` to be reused (if we want to allow that) by the implementations, although we'd have to make it public to support that use case. App servers would be able to create an instance per application, and pass it to the new `MetricID` constructor.

I've also marked old constructors as `@Deprecated`.

Reply:
I strongly disagree with deprecating the original two constructors. These are being used by application developers.
Why should we force application developers to use a constructor that takes a `GlobalTagsProvider`? That thing should only be concern of the implementation, not app developers.

I think a suitable solution would be that the `MetricID` constructors use the `ServiceLoader` to see if there are any custom `GlobalTagsProvider`s, and if yes, use them; if not, then use the default one. 

With the currently proposed code, if you use the `DefaultGlobalTagsProvider`, it will also locate and call the custom ones, which is not what one would expect to happen if using the default. Even worse, if your custom ones return different tags over time, the default provider will cache the result after first calling the custom ones, which is unexpected.
Either only the default should be used, or only the custom ones.

Reply:
I strongly disagree with deprecating the original two constructors. These are being used by application developers.
Why should we force application developers to use a constructor that takes a `GlobalTagsProvider`? That thing should only be concern of the implementation, not app developers.

I think a suitable solution would be that the `MetricID` constructors use the `ServiceLoader` to see if there are any custom `GlobalTagsProvider`s, and if yes, use them; if not, then use the default one. 

With the currently proposed code, if you use the `DefaultGlobalTagsProvider`, it will also locate and call the custom ones, which is not what one would expect to happen if using the default. Even worse, if your custom ones return different tags over time, the default provider will cache the result after first calling the custom ones, which is unexpected.
Either only the default should be used, or only the custom ones.

Reply:
> Why should we force application developers to use a constructor that takes a GlobalTagsProvider? That thing should only be concern of the implementation, not app developers

👍

Reply:
No strong feelings on deprecation, either works for me. I'll back out the change.

The rest of it is not so straight forward, though. 

I don't think passing `ServiceLoader` directly is the right thing to do. It really shouldn't be either/or (default or custom), because I'm not sure how the spec would even define that (each implementation would be free to implement behavior any way they want, which would hurt portability and defeat the purpose of having a spec in a first place). It is also inefficient, as it would suffer from the same issue current constructor has, which is lookup and calculation of **all** global tags every time new `MetricID` is constructed.

I think we need to refocus on defining the **spec** first, and then we can implement it accordingly.

1. Do we, or do we not want to initialize global tags only once? We seem to be going back and forth on this, based on discussion above? Yes, I understand that the current spec does not define that behavior, which caused the original complaint about current implementation's inefficiency, but I was under the assumptions that we all agreed that calculating a set of global tags once and caching them makes sense, for a variety of reasons. Let me know if that's not the case.

2. Do we want to provide a mechanism other than the config for spec **users** to provide global tags? This is where the notion of having `ServiceLoader`-based mechanism came from, and the reason I started looking into this in a first place, as we have a real need for that feature. If we want it, that behavior/feature needs to be defined by the spec.

3. If we have both `ServiceLoader` and config-based mechanism to provide global tags, what should the order/priority be? That's the reason I chose not to use `ServiceLoader`-based implementation for the config-based tags, as it allows us to control the ordering, but again, this needs to be defined by the spec.

So let's answer the questions above first, and once we do I'm happy to change the implementation accordingly.

Reply:
> Do we, or do we not want to initialize global tags only once? 

We do want to cache global tags - or at least allow the implementation to do that if they choose to - it shouldn't be mandatory.

> Do we want to provide a mechanism other than the config for spec users to provide global tags? 

I assume that by "spec users" you mean the MP Metrics implementation (not application developers), right? Of course we want to open ways for the implementation to do things as they want, in addition to the default behavior defined by the spec. That's why I think the `ServiceLoader` approach is a way to go.

On the other hand, we probably don't want application developers to have to be concerned about global tag providers. That's why I think we don't need a new `MetricID` constructor or deprecate the original ones. From the app developer point of view, defining the tags should be a matter of just configuring the server.

> If we have both ServiceLoader and config-based mechanism to provide global tags, what should the order/priority be? 

I think one problem here is that these two mechanisms you mention are not necessarily two completely separate things - a ServiceLoader approach may be *needed* to implement a customized config-based mechanism. 

If an implementation wants to do the MP Config-based mechanism (which they have to to pass the TCK), they might need to use a different piece of code for that than the code which we supply as default. Examples when this will be needed:
- an application server where the returned global tags depend on the application which calls the tag provider.
- or if it depends on some vendor-specific configuration mechanism other than reading the `mp.metrics.tags` property
So in this case they would need the ServiceLoader approach to plug in their own implementation of a config-based mechanism.

So what I think would be a solution that fits (I hope) all use cases:
- Provide a default `GlobalTagsProvider` that looks into MP Config **only once** to get the `mp.metrics.tags` property, generates tags based on that, caches them, and each call to it will return these cached tags. This should work for **most** single-app microservice runtimes, at least those that just stick to the spec only and don't provide any custom mechanisms on their own.

- Open a way to provide a custom `GlobalTagsProvider` through ServiceLoader, and if any such provider is present, this provider (providers) will be used **instead of** the default one. This will be for the implementations that need to do something specific, including considering multiple applications (looking at the `mp.metrics.appName` property, checking the TCCL or some thread-locals, or whatever else to determine the current application) to generate tags. Such custom provider would, of course, also need to take into account the regular `mp.metrics.tags` property to stay compliant.

- The `MetricID` constructor will always either call the default provider, or the ones from ServiceLoader. It won't cache the obtained results, potential caching will be responsibility of the providers (the default one will do it always, the custom ones can choose to do it).

- With this approach we would remain backward-compatible for most implementations, because we would be providing the default provider which reads the `mp.metrics.tags` property

- There remains a question whether the *implementation* (not application directly) might want to use a different `GlobalTagProvider` each time it is registering a metric. Because that would be a reason to allow passing it to the `MetricID` constructors. But I don't currently see any reasonable use case why an implementation would want to do this - if they want tags to be somehow "dynamic" (whatever that means exactly), these dynamicity can be implemented within the providers themselves rather than by switching between different providers.

I hope that clears up my views on this :)

Reply:
I agree with almost everything you said @jmartisk .  One thing I want to poke at is whether we truly need GlobalTagProvider at all.  Your two stated use cases are:

> an application server where the returned global tags depend on the application which calls the tag provider.

Here, Liberty has been able to just use MP Config to read application-specific versions of each tag.

> or if it depends on some vendor-specific configuration mechanism other than reading the mp.metrics.tags property

do we have a true request for allowing a config mechanism other than reading from MP Config?  I think we need clarity on why a component wouldn't just add its component-wide tags to the array of tags when creating a new metric.  If that truly is a need (component-wide tags) then I agree with need for GlobalTagProvider.

Reply:
> do we have a true request for allowing a config mechanism other than reading from MP Config?

We don't have a request here, but some implementations may choose to do introduce their own mechanism on top of the spec, for example WildFly has got its very specific configuration model (which is based neither on MP Config nor system properties), and they may choose to read global tags configured within that model as well

And determining the application name might also need some vendor-specific mechanism, it might not be just about reading `mp.metrics.appName`, especially if a multi-app server does not support separate configuration spaces per each application (or chooses to not use `mp.metrics.appName` at all, that's just a recommendation, not requirement). In such cases, the implementation may, for example, need to check the TCCL or some thread-local variables to be able to determine which application is the active one.